### PR TITLE
JIT: Improve codegen around unsigned comparisons

### PR DESCRIPTION
**Pull request by EgorBo**
1. Recognize more "known non-negative" idioms in VN
2. Normalize unsigned comparisons to signed when both of their operands are known non-negatives. Many existing optimizations such as RBO often give up on unsigned comparisons 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=757883&view=ms.vss-build-web.run-extensions-tab)

Motivated by https://github.com/dotnet/runtime/pull/104728#discussion_r1687893596

### CODE REVIEW

**Review by TIHan — APPROVED** [review]

**Comment by hez2010** [comment]
In `System.Globalization.TextInfo:ChangeCaseCommon[System.Globalization.TextInfo+ToLowerConversion](System.String):System.String:this (Instrumented Tier1)`, a new bound check appears which seems to be a regression.

**Comment by EgorBo** [comment]
> a new bound check appears which seems to be a regression.

Changes like these always produce a few regressions. In this case the problem that VN version of IsNeverNegative currently is not able to recognize Span.Length. It's something I want to fix one  day

**Comment by xtqqczze** [comment]
@MihuBot

**Comment by EgorBo** [comment]
an unrelated quick test:

@EgorBot -amd -arm64 -profiler
```cs
using System.Text.Json;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class MyBench
{
    public static IEnumerable<MyObj[]> TestData()
    {
        MyObj[] testData = new MyObj[100];
        for (int i = 0; i < testData.Length; i++)
        {
            MyObj obj1 = new("Some long ASCII text bla bla bla", 42, Guid.NewGuid(), DateTime.Now, null);
            MyObj obj2 = new("Какой-то non-ASCII text", i, Guid.NewGuid(), DateTime.Now, obj1);
            testData[i] = obj2;
        }
        yield return testData;
    }

    [Benchmark]
    [ArgumentsSource(nameof(TestData))]
    public object JsonStatham(MyObj[] arrays) => JsonSerializer.Serialize(arrays);
}

public record MyObj(string Name, int Age, Guid Id, DateTime Time, MyObj? InnerObj);
```

**Comment by EgorBo** [comment]
an unrelated quick test:

@EgorBot -amd -arm64 -profiler --envvars DOTNET_JitNoInline:1
```cs
using System.Text.Json;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

public class MyBench
{
    public static IEnumerable<MyObj[]> TestData()
    {
        MyObj[] testData = new MyObj[100];
        for (int i = 0; i < testData.Length; i++)
        {
            MyObj obj1 = new("Some long ASCII text bla bla bla", 42, Guid.NewGuid(), DateTime.Now, null);
            MyObj obj2 = new("Какой-то non-ASCII text", i, Guid.NewGuid(), DateTime.Now, obj1);
            testData[i] = obj2;
        }
        yield return testData;
    }

    [Benchmark]
    [ArgumentsSource(nameof(TestData))]
    public object JsonStatham(MyObj[] arrays) => JsonSerializer.Serialize(arrays);
}

public record MyObj(string Name, int Age, Guid Id, DateTime Time, MyObj? InnerObj);
```

**Comment by EgorBo** [comment]
PTAL @AndyAyersMS @BruceForstall @dotnet/jit-contrib 
small change, [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=826451&view=ms.vss-build-web.run-extensions-tab).

TLDR: if we have `op1 unsigned-relop op2` and we know that `op1` and `op2` are never negative - we can "normalize" the comparison to "signed-relop". Many existing branch/etc optimizations give up on unsigned relops due to, potentially, complex semantics.

**Review by AndyAyersMS — APPROVED** [review]
